refactoring vor v2 request profiles

mbaudis opened the pull request:
This PR creates a basic structure for request profiles, limited to the v2 variant parameters (i.e. not "VRSifying"). The main changes are:

* re-create `requestParameterComponents` in `common`
* create `requestProfiles` in common
* have the `g_variant` object in `genomicVariations/requestParameters/` being a `oneOf` of one of the request profiles

This PR does not yet include all/correct profiles and parameters but is used  for comparison and exploration of the potential structure. Changes need probably to be added before merging.

mbaudis committed:
start of refactoring for v2 request profiles

This first commit creates a basic structure for request profiles, limited to the v2 variant parameters (i.e. not "VRSifying". The main changes are:
* re-create `requestParameterComponents` in `common`
* create `requestProfiles` in common
* have the `g_variant` object in `genomicVariations/requestParameters/` being a `oneOf` of one of the request profiles

This commit does not yet include all/correct profiles and parameters but is for exploration of the potential structure.

mbaudis committed:
start of refactoring vor v2 request profiles

This first commit creates a basic structure for request profiles, limited to the v2 variant parameters (i.e. not "VRSifying". The main changes are:
* re-create `requestParameterComponents` in `common`
* create `requestProfiles` in common
* have the `g_variant` object in `genomicVariations/requestParameters/` being a `oneOf` of one of the request profiles

This commit does not yet include all/correct profiles and parameters but is for exploration of the potential structure.

mbaudis committed:
missed changes before PR...

mbaudis committed:
Merge branch 'develop-add-v2-request-types' of https://github.com/ga4gh-beacon/beacon-v2 into develop-add-v2-request-types

costero-e reviewed (commented):
Hi @mbaudis,

Thank you for the PR, I think it has great additions. Nonetheless, I always feel that `RefSeqId`, `Assembly` and now `VariantId` are very open, that the only restriction is for them to be a property of type `string`, and that makes them very diversely used, which is good, but then causes a problem of beacons converging for a same query. The beacons can have negative results for a query that means the same as another query but depending on how you do it you will get results or not, and I really dislike that. I solve this with a validator of queries for my implementation, but still, there are queries that are lost. It's not that I have come to a solution for this, maybe we could specify the nomenclatures we support to be more explicit and solve this, but I wanted to know what do you think about it.
Or maybe `VariantId` or another more explicit parameter can help to solve this.
Following this, I would make `RequestProfileId` to be an `enum` of the possible profiles accepted, more than just a `string`.

Let me know what you think about it.

PS: Typo in line 5 of `requestParameterComponents.json`: `refereenceBases`.

mbaudis commented:
> Thank you for the PR, I think it has great additions. Nonetheless, I always feel that `RefSeqId`, `Assembly` and now `VariantId` are very open, that the only restriction is for them to be a property of type `string`, and that makes them very diversely used, which is good, but then causes a problem of beacons converging for a same query.

Yes, this is a problem. But: This PR is explicitly to _refactor_ using *existing parameters* (with the addition of `profileId` - note changed name). Obviously we have to move to versioned sequence identifiers and drop the `assembly` altogether next, and do this in line with the VRS practices (refget compatible id, _i.e._ either `refseq:...` or `ga4gh:...`).

> t's not that I have come to a solution for this, maybe we could specify the nomenclatures we support to be more explicit and solve this, but I wanted to know what do you think about it. 

Again, true. And the procedural way here could be staggered interpretation, e.g. using seqrepo for checking:

* if not prefixed:id (`refseq:NC_000018.10`)
   - => try if assembly:referenceName (`GRCh18:18` or `GRCh18:chr18`) works.

... but this is done implementation side.

So the proposal here is as said above: A new, clearly defined parameter aligned with/ VRS which we've already prototyped for the next iteration: https://genomebeacons.org/variant-query-types/generated/requestParameterComponents/#refgetaccession ... done ASAP, together with other VRS/VQS parts, but not as part of this PR.

> Following this, I would make `RequestProfileId` to be an `enum` of the possible profiles accepted, more than just a `string`.

Not needed since `profileId` in the profiles is a `const` and you avoid double definitions (or rather triple - since we define `g_variant` request as `oneOf` ... in `.../genomicVariations/requestParameters`). So a profile check fetches the definition from the `profileId`, compares if the parameters are all there and constructs the query.

Added: The `additionalProperties: false` in `Bv2GenericVariantRequest` actually leads to the failure of requests with a wrong `profileId`; _i.e._ your requests either have a `profileId` which exists in one of the profiles, or there is no `profileId` in which case it becomes a generic v2 request. AFAIK `¯\_(ツ)_/¯`

> PS: Typo in line 5 of `requestParameterComponents.json`: `refereenceBases`.

Thanks - fixed.

mbaudis committed:
requestProfile -> profileId

* changes `requestProfile` -> `profileId`
* fixes typo `refereenceBases` (thanks @costero-e)

mbaudis commented:
And @costero-e , @jrambla : I'd be happy to add the `referenceAccession` => [RefgetAccession](https://genomebeacons.org/variant-query-types/generated/requestParameterComponents/#refgetaccession) directly here to have an unanimous reference identifier; but I thought the idea wasn't to have changes in substance already here?

costero-e commented:
Thank you for the extensive response @mbaudis, that was very well explained. I have no objections on the current PR and +1 to add an unanimous reference identifier, although maybe it was meant to implement it in the future, to come with the acceptance of VRS type parameters. Maybe @jrambla knows better.

mbaudis committed:
g_variant -> Bv2GenericVariantRequest

Renamed the generic variant request profile reference from `g_variant` to `Bv2GenericVariantRequest`; the profile itself had already that name

mbaudis committed:
profileId ==>> requestProfileId

This changes the proposed parameter from `profileId` to `requestProfileId`, to improve specificity and in line with the `RequestProfileId` prototype definition.

mbaudis committed:
Merge pull request #274 from ga4gh-beacon/develop

Update branch from develop

mbaudis committed:
Adding `requestProfileId` to `genomicVariations/endpoints`

Adding `requestProfileId` parameter to `genomicVariations/endpoints`
